src/charm.py

```diff
@@ -782,6 +779,25 @@ def _get_user_or_fail_event(self, event: ActionEvent, default_username: str) ->
             return
         return username
 
+    def pass_pre_set_password_checks(self, event: ActionEvent) -> bool:
+        """Checks conditions for setting the password and fail if necessary."""
+        if self.is_role(Config.Role.SHARD):
+            event.fail("Cannot set password on shard, please set password on config-server.")
+            return
+
+        # changing the backup password while a backup/restore is in progress can be disastrous
+        pbm_status = self.backups._get_pbm_status()
+        if isinstance(pbm_status, MaintenanceStatus):
+            event.fail("Cannot change password while a backup/restore is in progress.")
+            return
+
+        # only leader can write the new password into peer relation.
+        if not self.unit.is_leader():
```

Comment:
Yes it is possible, that the unit is both `leader` and `SHARD`. But it doesn't matter if the unit is `leader` if it is a `SHARD`, since as long as the unit is a `SHARD` we should not allow password rotation.